### ISSUE TITLE
Create public interfaces for all internal DOM classes

### ISSUE THREAD

**Issue by doominator42** [AngleSharp/AngleSharp#721]
Most of DOM element classes have an interface (e.g. HtmlAnchorElement : IHtmlAnchorElement), but some (e.g. `HtmlStrongElement`, `HtmlItalicElement`, `HtmlNoScriptElement`) does not have one. As those classes are internal, we can't identify which element it is (e.g. `element is IHtmlStrongElement`).

For now, I forked the project and will create the interfaces on my own.
I will create a pull request when done if this change is accepted.

**Comment by FlorianRappl** [comment]
Why do you want to distinguish on the class? What benefit does the interface give you (afaik there is no new functionality)? You can distinguish via, e.g., `TagName` or `NodeName`.

**Comment by doominator42** [comment]
> Why do you want to distinguish on the class?

I code a scraper. It consist of multiple scraper types for each DOM element types. When parsing DOM it calls the appropriate scraper based on element type.

> You can distinguish via, e.g., `TagName` or `NodeName`.

AngleSharp already does this classifying thing by creating appropriate objects for specific tag names.
Then it's natural to do `element is IHtmlSomeElement` rather than `string.Equals(element.TagName, TagNames.Some, StringComparison.OrdinalIgnoreCase)`.
Also some types have multiple associated tag names (e.g. HtmlSemanticElement), so I would have to hardcode a list of tag names for elements which is already done in AngleSharp.

And what's the purpose of having a specific element type if we can't do anything with it because it's internal and it doesn't have a public interface?

**Comment by FlorianRappl** [comment]
> AngleSharp already does this classifying thing by creating appropriate objects for specific tag names.
Then it's natural to do element is IHtmlSomeElement rather than string.Equals(element.TagName, TagNames.Some, StringComparison.OrdinalIgnoreCase).

These classes are internal because they may change - we only publish interfaces that contain functionality and are considered stable (i.e., by following the offical specs).

> Also some types have multiple associated tag names (e.g. HtmlSemanticElement), so I would have to hardcode a list of tag names for elements which is already done in AngleSharp.

This is true, but then you could do the same via reflection and getting the name of the base class. But this is a valid point.

> And what's the purpose of having a specific element type if we can't do anything with it because it's internal and it doesn't have a public interface?

See point above - in the earlier versions we published all the classes but this became a true show stopper (like other things now). Once its out there it needs to be maintained and compatibility for moving things is always an issue.

Your PR changes 93 files (imho way too many) and introduces an interface for every element, nearly all of them *empty* and some of them with functionality found in other interfaces. Also it does not even achieve your own goals (e.g., in the grouping area instead of `foo is InsertSingleInterface` you would still need to do `foo is A || foo is B || ...`).

TL;DR: I think the PR has a solid basis, but it needs to be reduced to the interfaces *at least* carrying functionality. If the interface is empty then there is no chance of inserting it.

**Comment by doominator42** [comment]
>Your PR changes 93 files (imho way too many)...

I don't really see the problem here... unless all those files are moved... like it seems to be in the `devel` branch. I'm aware that this will cause many conflicts. If I had to start from the `devel` branch, then I could create a new PR to that branch and consolidate changes.

>you would still need to do `foo is A || foo is B || ...`

That was an example of usage. In my project, there are dynamically registered classes inheriting `Scraper<TElement>`. There is also a list of excluded element types, which contains most of empty interfaces.

>I think the PR has a solid basis, but it needs to be reduced to the interfaces _at least_ carrying functionality. If the interface is empty then there is no chance of inserting it.

I think an empty interface still have a functional usage (the type itself). The class (e.g. `HtmlBoldElement`) doesn't have any members either, then why it's there? It's there because it conceptually represents an distinct element (or category of elements), despite the element doesn't have attributes, isn't that true?

My goal is to identify and process elements by using object-oriented patterns as this is how this library is designed, isn't it?

PS: I feel this is going nowhere. If those changes are unacceptable, then close the PR. I tought that would be usefull for others. Anyway I already integrated the new interfaces in my project, so I'll use them no matter happens.

**Comment by FlorianRappl** [comment]
> My goal is to identify and process elements by using object-oriented patterns as this is how this library is designed, isn't it?

Yes and no. Internally this may be the case, but from the outside it should / has to feel like an implementation of the official API / specs. This is one of the goals mentioned in the README.

> The class (e.g. HtmlBoldElement) doesn't have any members either, then why it's there? 

Good question - I cannot recall (it may be for future reuse, but other than that I cannot tell without doing some research - from the top of my head it was just a shorthand to ensure creating a bold element always has the right tag and flags - it transports this information to the base constructor). If its really not needed then we should eventually remove it or replace it with helper functions.

Yes, of course feel free to use it as you like (I mean that's the purpose of being open source). Your PR is definitely appreciated (just to avoid any other impression) I just don't feel it matches the overall ambition or goals (well, as said the empty ones I do not regard as useful, but some of the others are definitely worth integrating).

**Comment by doominator42** [comment]
>Yes and no. Internally this may be the case, but from the outside it should / has to feel like an implementation of the official API / specs. This is one of the goals mentioned in the README.

Ok, that makes more sense now. I forgot that AngleSharp is meant to be a browser, not only a parser.

I removed the empty interfaces for the rest to be merged.

Btw, there are some undocumented members in interfaces because there was no doc in the class and I don't really know the purpose of those members.

**Comment by FlorianRappl** [comment]
> Btw, there are some undocumented members in interfaces because there was no doc in the class and I don't really know the purpose of those members.

Don't forget that public members of an internal class are also internal - therefore these members are usually not for public use and therefore not documented. Their purpose should be (but exceptions can exist) obvious from their name + find reference / usage. Let me know which ones you think may need documentation / make sense to be available publicly, so that we can improve it!

Thanks for making the change 🍻 !
